### PR TITLE
Fix #658: Add possibility to configure error response behaviour.

### LINKED ISSUE #658: Client friendly requests blocking
Currently HTTP parser can return TFW_BLOCK in many cases. Meantime when TFW_BLOCK is returned from `tfw_http_msg_process()`, then the message is just silently dropped. So if a client uses poor browser, then they typically just see hang pages loading. System administrator should be able to specify the behaviour: return HTTP error message to a client of just silently block malformed and malicious message (requests and responses). This could be done by a new configuration option like

        block_action <drop|reply> [nolog]

`drop` is just current behaviour while `reply` sends error response to a client. Optional argument `nolog` requires don't do logging the reason for blocking. By default Tempesta FW must write to the log why particular message was blocked. I assume it could be done by replacing statements like 

        return TFW_BLOCK;

by

        return tfw_block("Content-Length with Transfer-Encoding in the same request");

, where `tfw_block()` checks configuration variable and does the logging.

### PR DESCRIPTION
**Pull request by aleksostapenko**

### CODE REVIEW

File: tempesta_fw/http.c [hunk 1]
```diff
@@ -713,6 +715,34 @@ tfw_http_req_error(TfwSrvConn *srv_conn, TfwHttpReq *req,
...
+	case 504:
+		tfw_http_send_504(req, reason);
+		break;
+	default:
+		TFW_WARN("Unexpected response error code: [%d]\n",
+			 status);
+		tfw_http_send_500(req, reason);
+		break;
+	}
+}
```
Comment (krizhanovsky) [inline tempesta_fw/http.c]: The comment https://github.com/tempesta-tech/tempesta/issues/658#issuecomment-279260358 isn't addressed. Please load files named like `404.html` or `500.html` from etc directory and use the content as a response body. Otherwise a user just sees blank pages. Currently we load TLS certificates from the files, so it'd be good to generalize the routines and use the same code for TLS certificates and the error responses. Probably a new variables, `$TFW_PATH/etc` should be passed to tempesta_fw.ko - the path can be used for TLS certificates as well as for the error codes.

File: tempesta_fw/sock_clnt.c [hunk 2]
```diff
@@ -37,15 +37,17 @@
  * ------------------------------------------------------------------------
  */
 
+unsigned short tfw_block_action_flags = TFW_BLOCK_ACTION_ERROR_REPLY;
```
Comment (krizhanovsky) [inline tempesta_fw/sock_clnt.c]: I'm wondering if we can define multiple rules

    block_action error reply;
    block_action attack drop nolog;

as described in https://github.com/tempesta-tech/tempesta/issues/658#issuecomment-270217864 using only this one variable? It seems the requirement is missed.

File: tempesta_fw/http.h [hunk 3]
```diff
@@ -290,6 +290,7 @@ typedef struct {
 /* URI has form http://authority/path, not just /path */
 #define TFW_HTTP_URI_FULL		0x000400
 #define TFW_HTTP_NON_IDEMP		0x000800
+#define TFW_HTTP_SUSPECTED		0x001000
```
Comment (keshonok) [inline tempesta_fw/http.h]: Like for `Conn_Suspected` (please see the comment below), I believe the name is misleading as the decision has been made already. Perhaps, simply `TFW_HTTP_MSG_BAD`? This name may mean either of a message being filtered out, or being malicious, or being erroneous.

File: tempesta_fw/http.c [hunk 4]
```diff
@@ -2159,7 +2241,9 @@ tfw_http_req_process(TfwConn *conn, struct sk_buff *skb, unsigned int off)
 
 		/* Assign the right Vhost for this request. */
 		if (tfw_http_req_set_context(req)) {
-			tfw_http_conn_msg_free((TfwHttpMsg *)req);
+			TFW_INC_STAT_BH(clnt.msgs_otherr);
+			tfw_http_drop(req, 500, "cannot find"
+				      "Vhost for request", true);
```
Comment (keshonok) [inline tempesta_fw/http.c]: It's been stated that special drop/block processing is for the cases where a request either malformed or malicious (i.e. filtered out). This particular case is more of an internal error nature (or maybe even a configuration). So why `tfw_http_drop()`? Why the connection has to be closed in this case? Does the client need to know of this error in the browser?

There are still regular `tfw_http_send_500()` and `tfw_http_send_502()` that weren't converted to this block/drop style, and that's justifiable and there's no need to block a request or drop a connection in those case.

File: tempesta_fw/http.c [hunk 5]
```diff
@@ -2216,8 +2300,9 @@ tfw_http_req_process(TfwConn *conn, struct sk_buff *skb, unsigned int off)
 				 */
 				TFW_WARN("Not enough memory to create"
 					 " a request sibling\n");
-				tfw_http_conn_msg_free((TfwHttpMsg *)req);
 				TFW_INC_STAT_BH(clnt.msgs_otherr);
+				tfw_http_drop(req, 500, "cannot create"
+					      " sibling request", true);
```
Comment (keshonok) [inline tempesta_fw/http.c]: Again, this seems like an internal error. Why the change to `tfw_http_drop()` Why the connection has to be closed in this case? Does the client need to know of this error in the browser?

File: tempesta_fw/http.c [hunk 6]
```diff
@@ -2058,6 +2073,69 @@ tfw_http_req_set_context(TfwHttpReq *req)
...
+/**
+ * Do the same as 'tfw_http_drop' function, but for malicious messages.
+ */
+static void
+tfw_http_block(TfwHttpReq *req, unsigned short code, const char *msg, bool mark)
+{
+	bool reply = tfw_block_action_flags & TFW_BLOCK_ACTION_ATTACK_REPLY;
+ 	bool nolog = tfw_block_action_flags & TFW_BLOCK_ACTION_ATTACK_NOLOG;
+	tfw_http_error_resp_and_log(reply, nolog, req, code, msg, mark);
+}
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply tempesta_fw/http.c]: @keshonok thank you for the notes!

The modern compiler doesn't treat `inline` C keyword actually. This is why we have attributes `noinline` and `always_inline`, so there is no big deal with inline specifiers in the patch. (However, I also don't like as they are made in the patch). Also, technically speaking, 6 arguments are OK - we have exactly 6 registers for arguments in Linux/x86-64 calling convention. Meantime, I agree that at least `mark` argument isn't a good one.

File: tempesta_fw/http.h (hunk 3 above)
Comment (krizhanovsky) [reply tempesta_fw/http.h]: I don't agree: `TFW_HTTP_MSG_BAD` doesn't say what `bad` is? Is it corrupted or malicious message? I'm OK with `suspected`.

Comment (keshonok) [reply tempesta_fw/http.h]: What I was trying to say was that in my opinion that there's nothing that is `suspected` at this point. The decision had been made, the message *IS* now either filtered out, malicious, or erroneous. Yet, all of these messages are now marked as suspected. Due to this decision the client connection will be closed.

So I didn't understand what it's suspected of?

Suspicion is when we don't know what it is yet, it's either this or that, and there are further actions to come up with more detailed knowledge.  But it doesn't work that way in the code. Again, the decision had been made at the time this flag is set, with the HTTP error response code and the error message. There's nothing that is suspected. Just the opposite, everything is pretty clear and definite.

Just a thought. :-)

File: tempesta_fw/http.c [hunk 7]
```diff
@@ -1789,7 +1802,9 @@ __tfw_http_resp_fwd(TfwCliConn *cli_conn, struct list_head *ret_queue)
 	list_for_each_entry_safe(req, tmp, ret_queue, msg.seq_list) {
 		BUG_ON(!req->resp);
 		tfw_http_resp_init_ss_flags((TfwHttpResp *)req->resp, req);
-		if (tfw_cli_conn_send(cli_conn, (TfwMsg *)req->resp)) {
+		if (tfw_cli_conn_send(cli_conn, (TfwMsg *)req->resp) ||
+		    (TFW_CONN_TYPE(cli_conn) & Conn_Suspected &&
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 8]
```diff
@@ -1789,7 +1802,9 @@ __tfw_http_resp_fwd(TfwCliConn *cli_conn, struct list_head *ret_queue)
...
+		     req->flags & TFW_HTTP_SUSPECTED)) {
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 9]
```diff
@@ -2058,6 +2073,69 @@ tfw_http_req_set_context(TfwHttpReq *req)
...
+			    unsigned short code, const char *msg, bool mark)
+{
+	TfwCliConn *cli_conn = (TfwCliConn *)req->conn;
+	if (reply) {
+		if (mark) {
+			BUG_ON(req->flags & TFW_HTTP_SUSPECTED ||
+			       TFW_CONN_TYPE(cli_conn) & Conn_Suspected);
+			TFW_CONN_TYPE(cli_conn) |= Conn_Suspected;
+			req->flags |= TFW_HTTP_SUSPECTED;
+			tfw_connection_unlink_msg(req->conn);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 10]
```diff
@@ -2058,6 +2073,69 @@ tfw_http_req_set_context(TfwHttpReq *req)
...
 
+/*
+ * Function defines logging and response behaviour during
+ * detection of malformed or malicious messages. Can be
+ * called from both - client and server connection contexts.
+ *
+ * NOTE: @mark must be set only for client connection context
+ * because in this case we must mark client connection in
+ * special manner to delay its closing until transmission
+ * of error response will be finished.
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected: added the ability to mark a request and connection for closing from the server context.

File: tempesta_fw/http.c [hunk 11]
```diff
@@ -2058,6 +2073,69 @@ tfw_http_req_set_context(TfwHttpReq *req)
...
+	else {
+		spin_lock(&cli_conn->seq_qlock);
+		if (unlikely(!list_empty(&req->msg.seq_list)))
+			list_del_init(&req->msg.seq_list);
+		spin_unlock(&cli_conn->seq_qlock);
+		tfw_http_conn_msg_free((TfwHttpMsg *)req);
+	}
+	if (!nolog)
+		TFW_WARN("Error response: %s, msg=%p conn=%p\n",
+			 msg, req, req->conn);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 12]
```diff
@@ -2058,6 +2073,69 @@ tfw_http_req_set_context(TfwHttpReq *req)
...
+
+/**
+ * Defines behaviour for malformed messages depending on configuration
+ * settings: sending response error messages and logging.
+ */
+static void
+tfw_http_drop(TfwHttpReq *req, unsigned short code, const char *msg, bool mark)
+{
+	bool reply = tfw_block_action_flags & TFW_BLOCK_ACTION_ERROR_REPLY;
+ 	bool nolog = tfw_block_action_flags & TFW_BLOCK_ACTION_ERROR_NOLOG;
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/sync_socket.h [hunk 13]
```diff
@@ -35,6 +35,12 @@ typedef struct ss_proto_t {
 	int			type;
 } SsProto;
 
+/* Flag bits offset for SsProto field. */
+#define	__Flag_Bits		0x10
+
+/* No requests longer accepted (flag only for client connections) */
+#define	Conn_Suspected		(0x1 << __Flag_Bits)
+
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/sync_socket.h]: Corrected: added comments and build-stage check for bit flags overlapping.

File: tempesta_fw/http.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: @keshonok thank you for review!

[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/sock.c [hunk 14]
```diff
@@ -858,10 +863,13 @@ ss_tcp_data_ready(struct sock *sk)
...
 	else if (!skb_queue_empty(&sk->sk_receive_queue)) {
-		if (ss_tcp_process_data(sk)) {
+		if (ss_tcp_process_data(sk) &&
+		    !(SS_CONN_TYPE(sk) & Conn_Suspected)) {
 			/*
 			 * Drop connection in case of internal errors,
-			 * banned packets, or FIN in the received packet.
+			 * banned packets, or FIN in the received packet,
+			 * and only if we don't wait for sending the error
+			 * response to client (Conn_Suspected bit set).
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/sock.c]: Corrected.

File: tempesta_fw/sync_socket.h [hunk 15]
```diff
@@ -35,6 +35,12 @@ typedef struct ss_proto_t {
 	int			type;
 } SsProto;
 
+/* Flag bits offset for SsProto field. */
+#define	__Flag_Bits		0x10
+
+/* No requests longer accepted (flag only for client connections) */
+#define	Conn_Suspected		(0x1 << __Flag_Bits)
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/sync_socket.h]: Corrected.

File: tempesta_fw/sock_clnt.c (hunk 2 above)
Comment (aleksostapenko) [reply tempesta_fw/sock_clnt.c]: The are 4 least significant bits to set/unset - so we have 16 possible states. And from the configuration statements:
> block_action error reply;
> block_action attack drop nolog;
we also have 16 possible combinations - so this should be enough.
If we have third directive. e.g.:
> block_action error reply;
then the corresponding flag bits will be rewritten to new values (if I understand the task correctly).

File: tempesta_fw/http.c (hunk 4 above)
Comment (aleksostapenko) [reply tempesta_fw/http.c]: > Does the client need to know of this error in the browser?
But in the opposite case - if we don't send error response - the client will see only hang pages.
Maybe it is more suitable for system administrator responsibility - to decide is it worth to send error responses in these cases.

File: tempesta_fw/http.c (hunk 1 above)
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected: configurable custom error pages added.

File: tempesta_fw/sock_clnt.c [hunk 16]
```diff
@@ -610,6 +680,13 @@ TfwCfgMod tfw_sock_clnt_cfg_mod  = {
 			.allow_repeat = false,
 			.cleanup = tfw_sock_clnt_cfg_cleanup_listen
 		},
+		{
+			"block_action",
+			NULL,
+			tfw_sock_clnt_cfg_handle_block_action,
+			.allow_repeat = true,
+			.allow_none = true
+		},
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/sock_clnt.c]: Done.

[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/sock_clnt.c]: Done.

File: tempesta_fw/sync_socket.h [hunk 17]
```diff
@@ -35,11 +35,23 @@ typedef struct ss_proto_t {
...
 
-/* No requests longer accepted (flag only for client connections) */
-#define	Conn_Suspected		(0x1 << __Flag_Bits)
+	/*
+	 * Connection is in special state: it is alive and
+	 * continue send responses to client, but no new
+	 * requests longer accepted (flag is intended
+	 * only for client connections).
+	 */
+	Conn_OnHold		= 0x1 << __Flag_Bits,
```
Comment (krizhanovsky) [inline tempesta_fw/sync_socket.h]: Well, I commented Aleksey's comment https://github.com/tempesta-tech/tempesta/pull/848#discussion_r144702374 and thought that it's enough and you just leave `suspected` name untouched. `OnHold` is bad name, at least because we already have `tfw_http_conn_on_hold()` which has completely different meaning. Moreover, the whole #658 change is about differentiation errors from security events (attacks), so all the related stuff must clearly define difference between the events instead of using common names.

Please revert the change.

File: tempesta_fw/http.c [hunk 18]
```diff
@@ -319,13 +433,17 @@ tfw_http_resp_build_error(TfwHttpReq *req)
...
- * NOTE: This function expects that the last chunk of @msg is CRLF.
+ * NOTE: This function expects the predefined order of chunks in @msg:
+ * the fourth chunk must be CRLF.
  */
-static void
-tfw_http_send_resp(TfwHttpReq *req, TfwStr *msg, const TfwStr *date)
+void
+tfw_http_send_resp(TfwHttpReq *req, resp_code_t code)
 {
+	TfwStr *msg = &http_predef_resps[code];
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: > Now you're working with static `http_predef_resps` variable and modify it, so it typically should crash on parallel error responses sending
Yes, I missed that :(

File: tempesta_fw/http.c [hunk 19]
```diff
@@ -68,6 +68,120 @@ int ghprio; /* GFSM hook priority. */
...
+	.len = 0,
+};
+
+static TfwStr http_5xx_resp_body = {
+	.ptr = (TfwStr []){
+		{ .ptr = NULL, .len = 0 },
+		{ .ptr = NULL, .len = 0 },
+	},
+	.len = 0,
+};
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Yes, I also thought about this approach. But in this case we have no the opportunity to set separately one page body for 500 answer and another page body for the remaining 5xx answers. As I understand it, the code 500 is a separate response code that does not intersect with the rest of the 5xx codes.

[the remark this answers is not in the document]
Comment (krizhanovsky) [reply tempesta_fw/http.c]: Yeah, makes sense. Just write a short comment why the two default static variables are required.

File: etc/tempesta_fw.conf [hunk 20]
```diff
@@ -344,6 +344,39 @@
...
+
+# TAG: response_body
+#
+# Syntax:
+#   response_body status_code file;
+#
+# Specifies a path to file with page body for defined HTTP status_code;
+# status_code must be present in three-digit form ('502', '403'), or
+# in form of codes group ('4*' or '5*').
+
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply etc/tempesta_fw.conf]: Done.

File: tempesta_fw/http.c [hunk 21]
```diff
@@ -68,6 +68,120 @@ int ghprio; /* GFSM hook priority. */
...
+	[RESP_200] = {
+		.ptr = (TfwStr []){
+			{ .ptr = S_200_PART_01, .len = SLEN(S_200_PART_01) },
+			{ .ptr = NULL, .len = SLEN(S_V_DATE) },
+			{ .ptr = S_200_PART_02, .len = SLEN(S_200_PART_02) },
+			{ .ptr = S_CRLF, .len = SLEN(S_CRLF) },
+			{ .ptr = NULL, .len = 0 },
+		},
+		.len = SLEN(S_200_PART_01 S_V_DATE S_200_PART_02 S_CRLF),
+		.flags = 4 << TFW_STR_CN_SHIFT
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Done.

File: tempesta_fw/http.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 22]
```diff
@@ -307,13 +433,17 @@ tfw_http_resp_build_error(TfwHttpReq *req)
...
+ * the fourth chunk must be CRLF.
  */
-static void
-tfw_http_send_resp(TfwHttpReq *req, TfwStr *msg, const TfwStr *date)
+void
+tfw_http_send_resp(TfwHttpReq *req, resp_code_t code)
 {
+	TfwStr *msg = &http_predef_resps[code];
+	TfwStr *date = __TFW_STR_CH(msg, 1);
+	TfwStr *crlf = __TFW_STR_CH(msg, 3);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 23]
```diff
@@ -2819,6 +2791,192 @@ tfw_http_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
...
+	size_t digs_count, l_size;
+	char *new_length, *new_body;
+	char buff[TFW_ULTOA_BUF_SIZ] = {0};
+
+	if (!(digs_count = tfw_ultoa(b_size, buff, TFW_ULTOA_BUF_SIZ))) {
+		TFW_ERR("too small buffer for Content-Length header\n");
+		return -E2BIG;
+	}
+
+	l_size = 2*SLEN(S_CRLF) + SLEN(S_F_CONTENT_LENGTH) + digs_count;
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 24]
```diff
@@ -2819,6 +2791,192 @@ tfw_http_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
...
+ * responses (for the cleanup case during shutdown).
+ */
+void
+tfw_http_del_resp_bodies(void)
+{
+	TfwStr *clen_str_4xx = __TFW_STR_CH(&http_4xx_resp_body, 0);
+	TfwStr *body_str_4xx = __TFW_STR_CH(&http_4xx_resp_body, 1);
+	TfwStr *clen_str_5xx = __TFW_STR_CH(&http_5xx_resp_body, 0);
+	TfwStr *body_str_5xx = __TFW_STR_CH(&http_5xx_resp_body, 1);
+	resp_code_t i;
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 25]
```diff
@@ -2819,6 +2791,192 @@ tfw_http_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
...
+	l_size = 2*SLEN(S_CRLF) + SLEN(S_F_CONTENT_LENGTH) + digs_count;
+	new_length = kmalloc(l_size + 1, GFP_KERNEL);
+	if (!new_length) {
+		TFW_ERR("can't allocate memory for Content-Length header\n");
+		return -ENOMEM;
+	}
+	snprintf(new_length, l_size + 1, "%s%s%s%s",
+		 S_CRLF, S_F_CONTENT_LENGTH , buff, S_CRLF);
+
+	new_body = kmalloc(b_size, GFP_KERNEL);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/http.c [hunk 26]
```diff
@@ -2819,6 +2791,192 @@ tfw_http_msg_process(void *conn, struct sk_buff *skb, unsigned int off)
...
+	TfwStr *clen_str_4xx = __TFW_STR_CH(&http_4xx_resp_body, 0);
+	TfwStr *body_str_4xx = __TFW_STR_CH(&http_4xx_resp_body, 1);
+	TfwStr *clen_str_5xx = __TFW_STR_CH(&http_5xx_resp_body, 0);
+	TfwStr *body_str_5xx = __TFW_STR_CH(&http_5xx_resp_body, 1);
+	resp_code_t i;
+	for (i = 0; i < RESP_NUM; ++i) {
+		TfwStr *body_str = __TFW_STR_CH(&http_predef_resps[i], 4);
+		if (body_str->ptr) {
+			TfwStr *clen_str =
+				__TFW_STR_CH(&http_predef_resps[i], 2);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/http.c]: Corrected.

File: tempesta_fw/sock_clnt.c [hunk 27]
```diff
@@ -655,6 +658,85 @@ tfw_sock_clnt_cfg_handle_block_action(TfwCfgSpec *cs, TfwCfgEntry *ce)
...
+	vfree(body_data);
+
+	return ret;
+}
+
+static void
+tfw_sock_clnt_cfg_cleanup_resp_body(TfwCfgSpec *cs)
+{
+	tfw_http_del_resp_bodies();
+}
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/sock_clnt.c]: Corrected.

File: tempesta_fw/sync_socket.h (hunk 17 above)
Comment (aleksostapenko) [reply tempesta_fw/sync_socket.h]: Corrected.

File: tempesta_fw/http.c [hunk 28]
```diff
@@ -19,6 +19,7 @@
  * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
 #include <linux/string.h>
+#include <linux/vmalloc.h>
```
Comment (krizhanovsky) [inline tempesta_fw/http.c]: You don't use `vmalloc()`, so it's better to use `<linux/gfp.h>` for `__alloc_free_pages()`

File: tempesta_fw/str.c [hunk 29]
```diff
@@ -271,6 +274,42 @@ tfw_strcpy(TfwStr *dst, const TfwStr *src)
...
+	if (unlikely(src_num != dst_num))
+		return -E2BIG;
+
+	c2 = dst_num ? dst->ptr : dst;
+	TFW_STR_FOR_EACH_CHUNK(c1, src, end) {
+		c2->ptr = c1->ptr;
+		c2->skb = c1->skb;
+		c2->eolen = c1->eolen;
+		c2->len = c1->len;
+		c2->flags = c1->flags;
```
Comment (keshonok) [inline tempesta_fw/str.c]: Perhaps, a simple `c2 = c1` would suffice here?

File: tempesta_fw/http.c (hunk 28 above)
Comment (aleksostapenko) [reply tempesta_fw/http.c]: But `vfree` function is used in `tfw_http_cfg_handle_resp_body` - to release the memory allocated in `tfw_cfg_read_file`.

File: tempesta_fw/str.c [hunk 30]
```diff
@@ -271,6 +274,42 @@ tfw_strcpy(TfwStr *dst, const TfwStr *src)
...
+	}
+	if (dst_num) {
+		/* The src and dst are compound. */
+		dst->len = src->len;
+		dst->flags = src->flags;
+	}
+
+	return 0;
+}
+EXPORT_SYMBOL(tfw_strcpy_desc);
```
[the remark this answers is not in the document]
Comment (aleksostapenko) [reply tempesta_fw/str.c]: Done.

File: tempesta_fw/str.c (hunk 29 above)
Comment (aleksostapenko) [reply tempesta_fw/str.c]: Yes, missed that.
Corrected.
